xl/player/gst/missing_plugin.py

```diff
@@ -0,0 +1,189 @@
+# exception to your version of the code, but you are not obligated to
+# do so. If you do not wish to do so, delete this exception statement
+# from your version.
+"""
+This module serves as frontend for GStreamer Plugin Base Utils (GstPbutils) for
+installing missing plugins.
+
+See also:
+* https://github.com/quodlibet/quodlibet/blob/master/quodlibet/quodlibet/player/gstbe/player.py#L548
+* https://git.gnome.org/browse/totem/tree/src/backend/bacon-video-widget-gst-missing-plugins.c
+
+This module must be a singleton, i.e. no instances are allowed. Thus, there is
+no class.
+
+Created on 25.04.2017
+
+@author: Christian Stadelmann
+"""
+
+import logging
+
+from gi.repository import GstPbutils
+
+from xl import event
+from xl.main import common
+from xl.nls import gettext as _
+
+LOGGER = logging.getLogger(__name__)
+
+MISSING_PLUGIN_URL = "https://exaile.readthedocs.io/en/stable/user/faq.html"
+
+# how to test:
+# the helper binary is located at /usr/libexec/gstreamer/gst-install-plugins-helper
+
+
+def handle_message(message, engine):
+    """
+        Handles `message` by checking whether it is a "missing plugin" message.
+        If it is, takes all required steps to make sure that
+            * playback is paused
+            * the user is notified of the issue
+            * the user gets a hint on which software to install
+
+        @param message: a Gst.Message of type Gst.MessageType.Element
+        @param engine: an instance of xl.player.gst.engine.ExaileGstEngine
+
+        @return: True if the message was a "missing plugin" message and was
+                    being handled. This does not mean that the plugin installed
+                    successfully.
+                 False if the message should be handled by some other code
+                    because it is not related to a missing plugin.
+    """
+    if not GstPbutils.is_missing_plugin_message(message):
+        return False
+
+    __handle_plugin_missing_message(message, engine)
+    return True
+
+
+def __handle_plugin_missing_message(message, engine):
+
+    desc = GstPbutils.missing_plugin_message_get_description(message)
+    installer_details = GstPbutils.missing_plugin_message_get_installer_detail(message)
+    LOGGER.warn("A plugin for %s is missing, stopping playback", desc)
+
+    user_message = _(
+        "A GStreamer 1.x plugin for %s is missing. "
+        "Without this software installed, Exaile will not be able to play the current file. "
+        "Please install the required software on your computer. See %s for details.") \
+        % (desc, MISSING_PLUGIN_URL)
+    # TODO make URL clickable by utilizing xlgui.widgets.dialogs.MessageBar
+
+    engine.stop()
+    __notify_user_on_error(user_message, engine)
+    if GstPbutils.install_plugins_supported():
+        if __run_installer_helper(installer_details):
+            return
+    LOGGER.warn("Installation of GStreamer plugins not supported on this platform.")
+
+
+@common.idle_add()
```

Comment:
Ok, I will change it. I didn't know that. Are there any rules on what is safe to do and what is not?